mat/band.go

```diff
@@ -0,0 +1,147 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mat
+
+import (
+	"gonum.org/v1/gonum/blas/blas64"
+)
+
+var (
+	bandDense *BandDense
+	_         Matrix  = bandDense
+	_         Band    = bandDense
+	_         RawBand = bandDense
+)
+
+// BandDense represents a band matrix in dense storage format.
+type BandDense struct {
+	mat blas64.Band
+}
+
+type Band interface {
+	Matrix
+	// Triangular returns the number of rows/columns in the matrix and its
+	// orientation.
+	Bandwidth() (kl, ku int)
+
+	// TBand is the equivalent of the T() method in the Matrix interface but
+	// guarantees the transpose is of band type.
+	TBand() Band
+}
+
+type RawBand interface {
+	RawBand() blas64.Band
+}
+
+var (
+	_ Matrix            = TransposeBand{}
+	_ Band              = TransposeBand{}
+	_ UntransposeBander = TransposeBand{}
+)
+
+// TransposeBand is a type for performing an implicit transpose of a Band
+// matrix. It implements the Band interface, returning values from the
+// transpose of the matrix within.
+type TransposeBand struct {
+	Band Band
+}
+
+// At returns the value of the element at row i and column j of the transposed
+// matrix, that is, row j and column i of the Band field.
+func (t TransposeBand) At(i, j int) float64 {
+	return t.Band.At(j, i)
+}
+
+// Dims returns the dimensions of the transposed matrix. Band matrices are
+// square and thus this is the same size as the original Triangular.
+func (t TransposeBand) Dims() (r, c int) {
+	c, r = t.Band.Dims()
+	return r, c
+}
+
+// T performs an implicit transpose by returning the Band field.
+func (t TransposeBand) T() Matrix {
+	return t.Band
+}
+
+// Bandwidth returns the number of rows/columns in the matrix and its orientation.
+func (t TransposeBand) Bandwidth() (kl, ku int) {
+	kl, ku = t.Band.Bandwidth()
+	return ku, kl
+}
+
+// TBand performs an implicit transpose by returning the Band field.
+func (t TransposeBand) TBand() Band {
+	return t.Band
+}
+
+// Untranspose returns the Band field.
+func (t TransposeBand) Untranspose() Matrix {
+	return t.Band
+}
+
+func (t TransposeBand) UntransposeBand() Band {
+	return t.Band
+}
+
+// NewBandDense creates a new Band matrix with r rows and c columns. If data == nil,
+// a new slice is allocated for the backing slice. If len(data) == r*(kl+ku+1),
+// data is used as the backing slice, and changes to the elements of the returned BandDense
+// will be reflected in data. If neither of these is true, NewBandDense will panic.
+//
+// The data must be arranged in row-major order. Only the values in the band portion
```

Comment:
I think this needs a definition of "row-major order", as if you haven't seen banded matrices it's not obvious what the layout is. Something like a copy and paste from the BLAS documentation.